File: app/src/main/java/com/reteno/sample/fragments/database/DeviceReadDialogFragment.java
package com.reteno.sample.fragments.database;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.transition.AutoTransition;
import androidx.transition.TransitionManager;

import com.reteno.core.data.local.database.manager.RetenoDatabaseManagerDevice;
import com.reteno.core.data.local.model.device.DeviceDb;
import com.reteno.sample.R;
import com.reteno.sample.databinding.ItemDbDeviceBinding;

import java.util.List;
import java.util.Locale;


public class DeviceReadDialogFragment extends BaseReadDialogFragment<DeviceDb, ItemDbDeviceBinding, DeviceReadDialogFragment.DeviceViewHolder, DeviceReadDialogFragment.DeviceAdapter> {

    private RetenoDatabaseManagerDevice databaseManager;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        databaseManager = serviceLocator.getRetenoDatabaseManagerDeviceProvider().get();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        databaseManager = null;
    }

    @Override
    protected void initAdapter() {
        adapter = new DeviceAdapter(() ->
                TransitionManager.beginDelayedTransition(bindingMain.rvList, new AutoTransition())
        );
    }

    @Override
    protected void initCount() {
        long deviceEventsCount = databaseManager.getDeviceCount();
        bindingMain.tvCount.setText(String.format(Locale.US, "Count: %d", deviceEventsCount));
    }

    @Override
    protected void deleteItems(int count) {
        databaseManager.deleteDevices(count, true);
    }

    @Override
    protected void initItems() {
        List<DeviceDb> newItems = databaseManager.getDevices(null);
        adapter.setItems(newItems);
    }

    //==============================================================================================
    static class DeviceAdapter extends BaseReadAdapter<DeviceDb, ItemDbDeviceBinding, DeviceViewHolder> {

        DeviceAdapter(ViewHolderListener listener) {
            super(listener);
        }

        @NonNull
        @Override
        public DeviceViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            ItemDbDeviceBinding binding = ItemDbDeviceBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
            return new DeviceViewHolder(binding);
        }

        @Override
        protected void initListeners(ItemDbDeviceBinding binding) {
            binding.ivExpand.setOnClickListener(v -> {
                if (binding.llContent.getVisibility() == View.VISIBLE) {
                    onExpandCollapseClickListener.onExpandCollapse();
                    binding.llContent.setVisibility(View.GONE);
                    binding.ivExpand.setImageResource(R.drawable.ic_expand_more);
                } else {
                    onExpandCollapseClickListener.onExpandCollapse();
                    binding.llContent.setVisibility(View.VISIBLE);
                    binding.ivExpand.setImageResource(R.drawable.ic_expand_less);
                }
            });
        }
    }

    //==============================================================================================
    static class DeviceViewHolder extends BaseReadViewHolder<DeviceDb, ItemDbDeviceBinding> {

        DeviceViewHolder(ItemDbDeviceBinding binding) {
            super(binding);
        }

        @Override
        protected void bind(DeviceDb device) {
            bindingHolder.tvDeviceId.setTextOrHide(device.getDeviceId());
            bindingHolder.tvExternalUserId.setTextOrHide(device.getExternalUserId());
            bindingHolder.tvPushToken.setTextOrHide(device.getPushToken());
            String pushSubscribed = null;
            if (device.getPushSubscribed() != null) {
                pushSubscribed = device.getPushSubscribed().toString();
            }
            bindingHolder.tvPushSubscribed.setTextOrHide(pushSubscribed);
            bindingHolder.tvCategory.setTextOrHide(device.getCategory().toString());
            bindingHolder.tvOsType.setTextOrHide(device.getOsType().toString());
            bindingHolder.tvOsVersion.setTextOrHide(device.getOsVersion());
            bindingHolder.tvDeviceModel.setTextOrHide(device.getDeviceModel());
            bindingHolder.tvAppVersion.setTextOrHide(device.getAppVersion());
            bindingHolder.tvLanguageCode.setTextOrHide(device.getLanguageCode());
            bindingHolder.tvTimeZone.setTextOrHide(device.getTimeZone());
            bindingHolder.tvAdvertisingId.setTextOrHide(device.getAdvertisingId());
        }
    }
}

File: app/src/main/java/com/reteno/sample/fragments/database/DeviceReadDialogFragment.kt
package com.reteno.sample.fragments.database

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.transition.AutoTransition
import androidx.transition.TransitionManager
import com.reteno.core.data.local.database.manager.RetenoDatabaseManagerDevice
import com.reteno.core.data.local.model.device.DeviceDb
import com.reteno.sample.R
import com.reteno.sample.databinding.ItemDbDeviceBinding
import com.reteno.sample.fragments.database.DeviceReadDialogFragment.DeviceAdapter
import com.reteno.sample.fragments.database.DeviceReadDialogFragment.DeviceViewHolder
import java.util.Locale

internal class DeviceReadDialogFragment :
    BaseReadDialogFragment<DeviceDb, ItemDbDeviceBinding, DeviceViewHolder, DeviceAdapter>() {

    private var databaseManager: RetenoDatabaseManagerDevice? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        databaseManager = serviceLocator.retenoDatabaseManagerDeviceProvider.get()
    }

    override fun onDestroy() {
        super.onDestroy()
        databaseManager = null
    }

    override fun initAdapter() {
        adapter = DeviceAdapter {
            TransitionManager.beginDelayedTransition(
                bindingMain!!.rvList, AutoTransition()
            )
        }
    }

    override fun initCount() {
        val deviceEventsCount = databaseManager!!.getUnSyncedDeviceCount()
        bindingMain!!.tvCount.text = String.format(Locale.US, "Count: %d", deviceEventsCount)
    }

    override fun deleteItems(count: Int) {
        val devices = databaseManager!!.getDevices(count)
        for (device in devices) {
            databaseManager!!.deleteDevice(device)
        }
    }

    override fun initItems() {
        val newItems = databaseManager!!.getDevices(null)
        adapter!!.setItems(newItems)
    }

    //==============================================================================================
    internal class DeviceAdapter(listener: ViewHolderListener) :
        BaseReadAdapter<DeviceDb, ItemDbDeviceBinding, DeviceViewHolder>(listener) {
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DeviceViewHolder {
            val binding =
                ItemDbDeviceBinding.inflate(LayoutInflater.from(parent.context), parent, false)
            return DeviceViewHolder(binding)
        }

        override fun initListeners(binding: ItemDbDeviceBinding) {
            binding.ivExpand.setOnClickListener {
                if (binding.llContent.visibility == View.VISIBLE) {
                    onExpandCollapseClickListener.onExpandCollapse()
                    binding.llContent.visibility = View.GONE
                    binding.ivExpand.setImageResource(R.drawable.ic_expand_more)
                } else {
                    onExpandCollapseClickListener.onExpandCollapse()
                    binding.llContent.visibility = View.VISIBLE
                    binding.ivExpand.setImageResource(R.drawable.ic_expand_less)
                }
            }
        }
    }

    //==============================================================================================
    internal class DeviceViewHolder(binding: ItemDbDeviceBinding) :
        BaseReadViewHolder<DeviceDb, ItemDbDeviceBinding>(binding) {
         override fun bind(model: DeviceDb) {
            bindingHolder.tvDeviceId.setTextOrHide(model.deviceId)
            bindingHolder.tvExternalUserId.setTextOrHide(model.externalUserId)
            bindingHolder.tvPushToken.setTextOrHide(model.pushToken)
            var pushSubscribed: String? = null
            if (model.pushSubscribed != null) {
                pushSubscribed = model.pushSubscribed.toString()
            }
            bindingHolder.tvPushSubscribed.setTextOrHide(pushSubscribed)
            bindingHolder.tvCategory.setTextOrHide(model.category.toString())
            bindingHolder.tvOsType.setTextOrHide(model.osType.toString())
            bindingHolder.tvOsVersion.setTextOrHide(model.osVersion)
            bindingHolder.tvDeviceModel.setTextOrHide(model.deviceModel)
            bindingHolder.tvAppVersion.setTextOrHide(model.appVersion)
            bindingHolder.tvLanguageCode.setTextOrHide(model.languageCode)
            bindingHolder.tvTimeZone.setTextOrHide(model.timeZone)
            bindingHolder.tvAdvertisingId.setTextOrHide(model.advertisingId)
        }
    }
}
